Add a new episode dealing with CMORizers and observational data (#178)
ESMValGroup/ESMValTool_Tutorial: _episodes/09-cmorization.md

> > long_name:         Carbon Mass in Vegetation Error
> > !----------------------------------
> > ! Additional variable information:
> > !----------------------------------
> > dimensions:        longitude latitude time
> > out_name:          cVegStderr
> > type:              real
> > !----------------------------------
> > !
> > ```
> >
> > Note that there is no entry for ``standard_name``. This is on purpose.
> > It is a sign for the ESMValTool to not crash although the variable that
> > we are looking for is ok to have no official CMIP6 ``standard_name``.
> >
> {: .solution}
{: .challenge}

## 5. Create a CMORizer for the new dataset

Now that we have the data ready, have told the ESMValTool where to look
for it, and have made sure that our variable of choice is listed either
on a pre-existing or custom CMOR table, let's test if the data is actually 
following the necessary CMOR standard already, or if we have to do some
reformatting for the dataset.

> ## Run the test recipe again
>
> Run the test recipe again from earlier where you wanted to check if the
> ESMValTool can already find and read the freshly downloaded FLUXCOM data.
>
> > ## Answers
> >
> > ```bash

> > > esmvaltool run recipe_check_fluxnet.yml --log_level debug

Author: I think it would still not find the data at this point since we have not yet run the cmorizer, right? So all we have is RAWOBS, whereas we need some data in OBS6 to get this error.

Maybe we still need to move the section "## Check if your variable is following the CMOR standard / Check if it's in a CMOR table" further down, so that the first thing we do is run the (almost) empty cmorizer script.

Reviewer: @Peter9192: yes, you are right. The ESMValTool would not find the data yet. It just did for me, I remember now, because I had it look in the RAWOBS folder...
And yes, I agree, we probably should have the "CMOR standard section" a little further down. Or at least have the people try to run an almost empty cmorizer script after the section "Location, location, location".

Reviewer: @Peter9192: actually, I have thought about a different approach, how to adjust our story line so that it makes more sense. I have implemented that now. Let me know what you think!

---

> title: "CMORization: Using observational datasets"

Reviewer 2: ```suggestion
title: "CMORization: using observational datasets"
```

Author: I'll change the title (and take care of the capitals)

Author: Done

teaching: 15
exercises: 45

questions:
- "What is so challenging about observational data?"
- "How do I use observational datasets in ESMValTool?"

> - "How add support for new (observational) datasets?"

Reviewer 2: ```suggestion
- "How do I add support for new (observational) datasets?"
```

Author: Changed to "how to add"

objectives:
- "Understand what CMORization is and why it is necessary."
- "Learn how to write a new CMORizer script."

keypoints:
- "CMORizers are dataset-specific scripts that can be run once to generate CMOR-compliant data."
- "ESMValTool comes with a set of CMORizers readily available, but you can also add your own."
---

## Introduction

> This episode deals with "CMORization". ESMValTool was designed to work with data

Reviewer 2: ```suggestion
This episode deals with "CMORization". ESMValTool is designed to work with data
```

Author: Done

that follow the CMOR standards. Unfortunately, not all datasets follow these
standards. In order to use such datasets in ESMValTool we first need to reformat
the data. This process is called "CMORization".

> ## What are the CMOR standards?
>
> The name "CMOR" originates from a tool: [the Climate Model Output
> Rewriter](https://cmor.llnl.gov/). This tool is used to create "CF-Compliant
> netCDF files for use in the CMIP projects". So CMOR extends the
> [CF-standard](https://cfconventions.org/) with additional requirements for
> the Coupled Model Intercomparison Projects (see e.g.
> [here](https://pcmdi.llnl.gov/CMIP6/Guide/modelers.html#5-model-output-requirements)).
>
>
> Concretely, the CMOR standards dictate e.g. the variable names and units,
coordinate information, how the data should be structured (e.g. 1 variable per
file), additional metadata requirements, but also file naming conventions a.k.a.
the data reference syntax (DRS). All this information is stored in so-called

> CMOR tables. As example, the CMOR tables for the CMIP6 project can be found

Reviewer 2: ```suggestion
CMOR tables. As an example, the CMOR tables for the CMIP6 project can be found
```

Author: Done

[here](https://github.com/PCMDI/cmip6-cmor-tables).
{: .callout}

The Earth System Grid Federation (ESGF) is home to all CMIP data. The data
hosted there typically (mostly) follow the CMIP standards, and therefore
ESMValTool should work with these data without problems.

Datasets that are *not* part of one of the CMIP projects often don't follow the
CMOR standards. In this case, a reformatting script can be used to create a
CMOR-compliant copy of these datasets. CMORizer scripts for several popular
datasets are included in ESMValTool, and ESMValTool also provides a convenient
way to execute them.

> Occasionally it happens that there are still minor issue with CMIP datasets. In

Reviewer 2: ```suggestion
Occasionally it happens that there are still minor issues with CMIP datasets. In
```

Author: I removed this sentence altogether

those cases, it is possible to fix those issues in ESMValCore before any further
processing is done. The same can be done for non-CMIP data. The advantage is
that you don't need to store an additional, reformatted copy of the data. The

> disadvantage is that these fixes should be implemented inside ESMValCore.

Reviewer 2: Please add a note here that this tutorial does not cover the development or any lessons about how to use esmvalcore.

Writing a CMORizer script is technically is simpler.

The concepts discussed so far are illustrated in the figure below.
![Data flow with ESMValTool](../fig/data_flow.png)
*Illustration of the data flow in ESMValTool.*

In this lesson, we will re-implement a CMORizer script for the FLUXCOM dataset
that contains observations of the Gross Primary Production (GPP), a variable
that is important for calculating components of the global carbon cycle. We will
go through all the steps and explain relevant topics as we go. If you prefer to
implement CMOR fixes, please read the documentation
[here](https://docs.esmvaltool.org/projects/esmvalcore/en/latest/develop/fixing_data.html#fixing-data).

> While fixes are implemented slightly differently, conceptually the process is

Reviewer 2: Please re-phrase this sentence regarding the usage of these three words "slightly differently, conceptually". The readability level of this sentence is quite low.

Author: Rephrased completely

the same and the concepts explained in this episode are still useful.

## Obtaining the data

> ## Get the data
> The data for this episode is available via the [FluxCom Data
> Portal](http://www.bgc-jena.mpg.de/geodb/BGI/Home). First you'll need to
> register. After registration, in the dropdown boxes, select FLUXCOM as the data
> choice and click download. Three files will be displayed. Click the download
> button on the "FLUXCOM (RS+METEO) Global Land Carbon Fluxes using CRUNCEP

> > climate data". You'll be send an FTP address to access the server. Connect

Reviewer 2: ```suggestion
> climate data". You'll receive an FTP address to access the server. Connect
```

Author: Done

> to the server, follow the path in your email, and look for the file

> > `raw/monthly/GPP.ANN.CRUNCEPv6.monthly.2000.nc`. Download that file and save in

Reviewer 2: ```suggestion
> `raw/monthly/GPP.ANN.CRUNCEPv6.monthly.2000.nc`. Download that file and save
```

Author: Corrected

> > in a folder called `/RAWOBS/Tier3/`.

Reviewer 2: The `/RAWOBS/Tier3/` is a directory rather than a folder name here. Can we make it clear by re-phrasing it?

Author: Not sure I understand. What's the difference between a directory and a folder name?

Reviewer 2: Because the folder name is `FLUXCOM` here, I thought the sentence “save it in a folder called `/RAWOBS/Tier3/FLUXCOM` " might be confusing from a user’s  (researcher) perspective as `/RAWOBS/Tier3/FLUXCOM ` shows three folders. However, from a developer’s perspective, I am aware that a folder is identified in this way. I see several cases in the lesson that a file or a folder is called in this way. So, it is fine with me.

>
> Note: you'll need a user-friendly ftp client. On Linux, `ncftp` works okay.

> {: .challenge}

Reviewer 2: Why the instruction about getting the data should be in a {:challenge} box while no solution is provided for this challenge? Also, we first teach something and then design a challenge. So users can practice what they have already learned. I suggest making the instruction plain text and remove one of these titles `Obtaining the data` and `Get the data`.

Author: Sorry I thought the challenge was also meant to distinguish exercise from theory. I've changed it to plain text.

> has created a tier-system. The definition of the different tiers are
> as follows:
>
> - **Tier1**: obs4MIPs and ana4MIPS datasets (can be used directly with the ESMValTool)
> - **Tier2**: other freely available datasets (most of them will need some kind of
>   cmorization)
> - **Tier3**: datasets with access restrictions (most of these datasets will also
>   need some kind of cmorization)
>
> These access restrictions are also the reason why the ESMValTool developers
> cannot distribute copies or automate downloading of all observations and
> reanalysis data used in the recipes. As a compromise we provide the
> CMORization scripts so that each user can CMORize their own copy of the access
> restricted datasets if they need them.
>
{: .callout}

## Make a test recipe

Now that we have downloaded the data, our end goal is to run a recipe that can
load it and work with it. So we will make this recipe and try to run it. It will
probably fail, but that's okay. As you will see, the error messages will come in
handy. Moreover, the recipe will serve as a test case. Once we get it to run, we
know we have completed our task.

> ## Create a test recipe
>

> > Create a simple recipe that loads the "FLUXCOM" data. It should include a

Reviewer 2: ```suggestion
> Create a simple recipe name `recipe_check_fluxnet.yml ` that loads the "FLUXCOM" data. It should include a
```

Author: Done

> datasets section with a single entry for the "FLUXCOM" dataset with the correct
> dataset keys, and a diagnostics section with two variables: gpp and gppStderr.

> > We won't need any preprocessors or scripts (set `scripts: null`), but you will

Reviewer 2: ```suggestion
> We don't need any preprocessors or scripts (set `scripts: null`), but we
```

Author: done

> have to add a documentation section with a description, authors and
> maintainer, otherwise the recipe will fail.
>
> Use the following dataset keys:
>
> - project: OBS6
> - dataset: FLUXCOM
> - type: reanaly
> - version: ANN-v1
> - mip: Lmon
> - start_year: 2000
> - end_year: 2000
> - tier: 3
>
> Some of these dataset keys are further explained in the callout boxes in this
> episode.
>
> > ## Answer
> >
> > Here's an example recipe
> >
> > ```yaml
> > documentation:
> >
> >   description: Test recipe for FLUXCOM data
> >
> >   authors:
> >     - kalverla_peter
> >
> >   maintainer:
> >     - kalverla_peter
> >
> > datasets:
> >   - {project: OBS6, dataset: FLUXCOM, mip: Lmon, tier: 3, start_year: 2000, end_year: 2000, type: reanaly, version: ANN-v1}
> >
> > diagnostics:
> >   FLUXCOM:
> >     description: Check that ESMValTool can load the cmorized FLUXCOM data without errors.
> >     variables:
> >       gpp:
> >       gppStderr:
> >     scripts: null
> >
> > ```
> >
> > Note: a recipe similar to this one is available under
> > `~/path/to/ESMValTool/esmvaltool/recipes/examples/recipe_check_obs.yml`.

> > > That recipe includes checks all datasets for which CMORizers are available.

Reviewer 2: I suggest replacing the path with its link e.g. [recipe_check_obs ](https://github.com/ESMValGroup/ESMValTool/blob/master/esmvaltool/recipes/examples/recipe_check_obs.yml.). Why users need to know about this recipe? Can we provide more information by referring users to some ESMValTool documentations? 

Another issue, we can add a tip here guiding users to have a look at the episode `Writing your worn recipe`.

Author: Removed the sentence with recipe_check_obs, and added the link to the episode on writing your own recipe 👍

> >
> {: .solution}
{: .challenge}

Try to run the example recipe with

```bash
esmvaltool run recipe_check_fluxcom.yml --log_level debug
```

The `log_level` flag ensures that all relevant information is included in the
output. We will need this information to find out what needs to be done. Look
carefully through the log messages. You'll probably find something like

```
DEBUG   Retrieving OBS6 configuration
DEBUG   Skipping non-existent /home/peter/default_inputpath/Tier3/FLUXCOM
DEBUG   Looking for files matching ['OBS6_FLUXCOM_reanaly_ANN-v1_Lmon_gppStderr[_.]*nc'] in []
ERROR   No input files found for variable {'variable_group': 'gppStderr' ...
...
ERROR   Missing data for preprocessor FLUXCOM/gppStderr
...
ERROR   Program terminated abnormally, see stack trace below for more information:
...
```
{: .error}

So the output tells us that it cannot find the data, and also where it has been
looking. Let's try to understand what's going on.

## Location, location, location

The error messages above tell you that ESMValTool cannot find the data. That

> makes total sense, because we have not yet created our CMORized copy of the

Reviewer 2: Sorry, I am lost here. I expect this episode shows me how to create a cmorizer script. There are already three well-written episodes on Configuration (episode 3), writing your own recipe (episode 5) and debugging (episode 6). Please avoided duplications and just point the users to those lessons. I see that the debugging exercise here is a nice one, please add it to episode 6.

Author: The idea was that the debug messages tell us what we need to fix for the cmorizer. How else would you know what kind of fixes the cmorizer needs to implement? But I can see why you're confused, and I think I know a way to avoid this here.

Author: I removed this, but the episode still uses the error messages in a few places. Is it okay like this, or do you still think it's confusing?

Reviewer 2: It is much clear now. Thank you for addressing my comment.

data. Our data is located in the `RAWOBS` folder, but ESMValTool is looking in
`OBS6`. So the first thing our CMORizer will need to do is copy the data over
from one folder to the other. To do end, we need to tell ESMValTool where our
data may be found.

> ## Set the correct paths in your user configuration file:
>
> This information is set in `config-user.yml`. Modify your
> configuration file so that it has the correct paths
>
> ```yaml
> rootpath:
>   OBS6: /path/to/my/obs6/data
>   RAWOBS: /path/to/my/rawobs/data
> ```

> {: .challenge}

Reviewer 2: Why the instruction about setting the correct path is in a {:challenge} box while no solution is provided for this challenge? Also, we first teach something and then design a challenge. So users can practice what they have already learned. I suggest making the instruction plain text.

Author: Sorry I thought the challenge was also meant to distinguish exercise from theory. I've changed it to plain text.

Reviewer 2: It is right that a challenge is "meant to distinguish exercise from theory". However, setting the correct path (here) and downloading data (in the above) are more **instructions** rather than **exercises**. 
A challenge provides learners an opportunity to practice what that you’ve included in the skill list (objectives) for designing the lesson. In designing a challenge, we have to first teach something then challenge learners if they can do it. I hope now it is clear why I asked to convert those instructions to plain texts.

> observations.
>
> Then, in going from CMIP5 to CMIP6, the CMOR standards changed a bit. Some
> variables are named differently in CMIP5 and CMIP6. This posed a dilemma:
> should CMORization reformat to the CMIP5 or CMIP6 definition? To solve this,
> the `OBS6` project was created. So data in the `OBS6` follow the CMIP6
> standards.
>
{: .callout}

Looking closer at the path that ESMValTool told us it went looking for the data,
we see that it has a very specific structure:

`OBS6_FLUXCOM_reanaly_ANN-v1_Lmon_gppStderr[_.]*nc`

Note that all components in the filename correspond to one of the dataset keys
defined in our test recipe. By replacing them with the original tags, we can
see the general structure for the OBS6 data:

`[project]_[dataset]_[type]_[version]_[mip]_[short_name]_*.nc`

This is the Data Reference Syntax (DRS) for OBS6 data. Each project has its own
DRS, and some project IDs even have multiple different options. These settings
can be found in the file
[config-developer.yml](https://github.com/ESMValGroup/ESMValCore/blob/master/esmvalcore/config-developer.yml).
Luckily, we don't need to worry about that. This file name will be automatically

> correctly created when we run our CMORizer script.

Reviewer 2: In this part, I am introduced to an example of DRS and then to the definition of DRS. Can we reverse the order i.e. first explain what a DRS is and then show me an example and then a reference to config-developer? 

Another point, in episode **configuration**, there is a section **Directory structure for the data from different projects** that explains DRS. Can we point users to that section?

Author: I removed most of this.

Most variables that we would want to use with the ESMValTool are defined in the Coupled
Model Intercomparison Project (CMIP) data request and can be found in the
CMOR tables in the folder
[cmip6/Tables](https://github.com/ESMValGroup/ESMValCore/tree/master/esmvalcore/cmor/tables/cmip6/Tables),
differentiated according to the "MIP" they belong to. The tables are a
copy of the [PCMDI](https://github.com/PCMDI) guidelines.

> ## Find the variable "gpp" in a CMOR table
>
> Check the available CMOR tables to find the variable "gpp" with the following characteristics:
> - standard_name: ``gross_primary_productivity_of_biomass_expressed_as_carbon``
> - frequency: ``mon``
> - modeling_realm: ``land``
>
> > ## Answers
> >
> > The variable "gpp" belongs to the land variables. The temporal resolution that we are looking
> > for is "monthly". This information points to the "Lmon" CMIP table. And indeed, the variable
> > "gpp" can be found in the file
> > [here](https://github.com/ESMValGroup/ESMValCore/blob/master/esmvalcore/cmor/tables/cmip6/Tables/CMIP6_Lmon.json).
> >
> {: .solution}
{: .challenge}

If the variable you are interested in is not available in the standard CMOR tables,

> you need to write a custom CMOR table entry for the variable. Don't worry! It sounds
> more complicated than it is! Examples of custom CMOR table entries are for example

Reviewer 2: ```suggestion
you need to write a custom CMOR table entry for the variable. Examples of custom CMOR table entries are for example
```

Author: This is no longer part of the episode

> > !----------------------------------
> > dimensions:        longitude latitude time
> > out_name:          cVegStderr
> > type:              real
> > !----------------------------------
> > !
> > ```
> >
> > Note that there is no entry for ``standard_name``. This is on purpose.
> > It is a sign for the ESMValTool to not crash although the variable that
> > we are looking for is ok to have no official CMIP6 ``standard_name``.
> >
> {: .solution}
{: .challenge}

## Create a CMORizer for the new dataset

Now that we have the data ready, our first assumption would be that the data
is already following the CMOR standard. So we store the file in the "OBS6"
folder of our data path, and rename it manually to an ESMValTool readable
name. For our example of the "FLUXCOM" data that we downloaded (which is
from the year 2000) the name of the dataset would then be:

```bash
OBS6_FLUXCOM_reanaly_ANN-v1_Lmon_gpp_200001-200012.nc
```

We also have made sure that our variable of choice is listed either
on a pre-existing or custom CMOR table. So now we are ready to test
if the data is actually following the necessary CMOR standard already,
or if we have to do some reformatting for the dataset.

> > ## Run the test recipe again

Reviewer 2: These sections are not shown correctly on the tutorial webpage. Please preview the lesson locally and fix it.

> >           method: Artificial Neural Networks
> >           provided_by: Martin Jung [mjung@bgc-jena.mpg.de] on behalf of FLUXCOM team
> >           reference: Jung et al. 2016, Nature; Tramontana et al. 2016, Biogeosciences
> >           source_file: /mnt/lustre02/work/bd0854/b309143/Data/Tier3/FLUXCOM/OBS_FLUXCOM_reana...
> >           temporal_resolution: monthly
> >           title: GPP based on FLUXCOM RS+METEO with CRUNCEPv6 climate
> >           version: v1
> > ...
> > ```
> {: .solution}
{: .challenge}

The error messages that we see tell us that we do have to reformat the dataset
slightly to have it follow the CMOR standard. It seems like the variable "gpp"
does not have the correct unit that is given in the CMOR table where it is
[defined](https://github.com/ESMValGroup/ESMValCore/tree/master/esmvalcore/cmor/tables/cmip6/Tables).
It also seems like the "latitude" and "longitude" coordinates have some
problems. So let's start writing a short python script that will fix these
problems.

The first step now is to create a file in the right folder that will contain
the short python script. The home of all CMORizer scripts for observations
and reanalysis datasets is

> [here](https://github.com/ESMValGroup/ESMValTool/tree/master/esmvaltool/cmorizers/obs).

Reviewer 2: I suggest replacing `[here] `to `esmvaltool/cmorizers/obs` and point users to episode development-setup where they can ding these directories.

Author: Good point. I'll instead add a note about the development setup episode near the top of the episode.

Author: I added several references to the dev setup episode, and I now point to these paths explicitly

Add a file with the name ``cmorize_obs_fluxcom.py`` to this folder.

> ## Note
>
> Always, always, when modifying or creating new code for the ESMValTool

> > repositories, work on your *own, local* branch of the ESMValTool. Optimally,
> > you have forked that branch directly from the most up-to-date version of
> > the "master" branch to avoid conflicts later when you want to merge your
> > code with the "master" branch of the ESMValTool.

Reviewer 2: ```suggestion
> repositories, work on your *own, local* branch of the ESMValTool. For more information see lesson [Development and contribution]({{ page.root }}{% link _episodes/08-development-setup.md %}).
```

Author: Pointing to the dev episode now

>
{: .callout}

Within our python script we have to make sure that we cover three main
aspects of the reformatting:
- We need to be able to locate and read the data to be CMORized
- We have to CMORize the data (fix the CMOR problems that the ESMValTool
had so nicely pointed out for us)
- We need to store the data with the correct filename so that the ESMValTool
will be able to identify it later

But the very first part of the CMORizing script is a header. The header
contains information about where to obtain the data, when it was accessed
the last time, which ESMValTool "tier" it is associated with, and more
detailed information about the necessary downloading and processing steps.

> > ## Fill out the header for the "FLUXCOM" dataset

Reviewer 2: These sections are not shown correctly on the tutorial webpage. Please preview the lesson locally and fix it.

> >     This constraint will not exist once iris is updated to
> >     version 2.3.0 Aug 2019
> > """
> > ```
> >
> > This is the header of the "FLUXCOM" CMORizer that is available with the
> > ESMValTool already. It is therefore entirely possible that your entries for
> > the section "Last access" and "Download and processing instructions"
> > differ from the example here since the entries for these sections are
> > somewhat subjective.
> >
> {: .solution}
{: .challenge}

Now that we have made sure, that it is clear where our dataset comes from, and
how we can obtain it, the next step is to add the python code to actually read
the data.

As we learned earlier, the DRS of the ESMValTool needs some very specific
pieces of information to build the name for each data file. Since we are now
trying to introduce a new dataset to the ESMValTool, we will have to specify
some things to make sure the dataset name can be correctly created, and
ultimately the ESMValTool knows how to look for the new dataset.

Therefore it is necessary to create a configuration file for the new dataset.
This configuration file needs to be stored in the

> following folder:
> 
> ```bash
> ESMValTool/esmvaltool/cmorizers/obs/cmor_config/
> ```

Reviewer 2: ```suggestion
following folder: ``ESMValTool/esmvaltool/cmorizers/obs/cmor_config/``
```

Author: No long part of the episode

> the ESMValTool framework. Try to fill in all missing pieces of information
> for this configuration file that are marked with ``???``.
>
> ```yaml
> ---
> # Filename
> filename: ???
>
> # Common global attributes for Cmorizer output
> attributes:
>   dataset_id: ???
>   version: ???
>   tier: ???
>   modeling_realm: ???
>   project_id: ???
>   source: ???
>   reference: ???
>   comment: ???
>
> # Variables to cmorize
> variables:
>   ???:
>     mip: ???
> ```
>
> > ## Answers
> >
> > The configuration file for the "FLUXCOM" dataset with all necessary pieces
> > of information looks  like this:
> >
> > ```yaml
> > ---
> > # Filename
> > filename: 'GPP.ANN.CRUNCEPv6.monthly.*.nc'
> >
> > # Common global attributes for Cmorizer output
> > attributes:
> >   dataset_id: FLUXCOM
> >   version: 'ANN-v1'
> >   tier: 3
> >   modeling_realm: reanaly
> >   project_id: OBS
> >   source: 'http://www.bgc-jena.mpg.de/geodb/BGI/Home'
> >   reference: 'fluxcom'
> >   comment: ''
> >
> > # Variables to cmorize
> > variables:
> >   gpp:
> >     mip: Lmon
> > ```
> >
> > The original configuration file for the "FLUXCOM" dataset can be
> > found here:

> > > [FLUXCOM.yml]
> > > (https://github.com/ESMValGroup/ESMValTool/blob/master/esmvaltool/cmorizers/obs/cmor_config/FLUXCOM.yml)

Reviewer 2: ```suggestion
> > [FLUXCOM.yml](https://github.com/ESMValGroup/ESMValTool/blob/master/esmvaltool/cmorizers/obs/cmor_config/FLUXCOM.yml)
```

Author: Done

> >
> > Note the attribute "reference" here: it should include a ``doi`` related
> > to the dataset. For more information on how to add references to the
> > ``reference`` section of the configuration file, see the section in the

> > > documentation about this: [adding references]
> > > (https://docs.esmvaltool.org/en/latest/community/diagnostic.html#adding-references)

Reviewer 2: ```suggestion
> > documentation about this: [adding references](https://docs.esmvaltool.org/en/latest/community/diagnostic.html#adding-references)
```

Author: No longer part of the episode

> >
> > If a single dataset has more than one reference, it is possible to add

> > > tags as a list e.g. ``reference: ['tag1', 'tag2']``.

Reviewer 2: This information is already mentioned in the documentation and it beyond the scope of this episode. Let's remove it to make the lesson concise.

Author: Done

> {: .solution}
{: .challenge}

Now that we have defined the configuration file for our "FLUXCOM" data, we can
finally start writing the actual code for the CMORizer script. The main body
of the CMORizer script must contain a function called

```python
def cmorization(in_dir, out_dir, cfg, config_user):
```

with this exact call signature. Here, ``in_dir`` corresponds to the input
directory of the raw files, ``out_dir`` to the output directory of final
reformatted data set and ``cfg`` to the configuration dictionary given by
the  ``.yml`` configuration file. The return value of this function is ignored.
All the work, i.e. loading of the raw files, processing them and saving the
final output, has to be performed inside its body. To simplify this process,

> ESMValTool provides a set of predefined utilities.py_, which can be imported

Reviewer 2: ```suggestion
ESMValTool provides a set of functions in ``utilities.py``, which can be imported
```

Author: No longer present

into your CMORizer by

```python
from . import utilities as utils
```

Apart from a function to easily save data, this module contains different
kinds of small fixes to the data attributes, coordinates, and metadata which
are necessary for the data field to be CMOR-compliant. We will come back to
these functionalities in a bit.

Note that this specific CMORizer script contains several subroutines in order
to make the code clearer and more readable (we strongly recommend to follow
that code style). For example, the function ``_get_filepath`` converts the raw
filepath to the correct one and the function ``_extract_variable`` extracts and
saves a single variable from the raw data.

> After all that theory, let's have a look at the actualy python code of the

Reviewer 2: ```suggestion
After all that theory, let's have a look at the python code of the
```

Author: No longer part of the episode

def _extract_variable(cmor_info, attrs, filepath, out_dir):
    """Extract variable."""
    var = cmor_info.short_name
    logger.info("Var is %s", var)
    cubes = iris.load(filepath)
    for cube in cubes:
        logger.info("Saving file")
        utils.save_variable(cube,
                            var,
                            out_dir,
                            attrs,
                            unlimited_dimensions=['time'])

def cmorization(in_dir, out_dir, cfg, _):
    """Cmorization func call."""
    glob_attrs = cfg['attributes']
    cmor_table = cfg['cmor_table']
    filepath = _get_filepath(in_dir, cfg['filename'])
    logger.info("Found input file '%s'", filepath)

    # Run the cmorization
    for (var, var_info) in cfg['variables'].items():
        logger.info("CMORizing variable '%s'", var)
        glob_attrs['mip'] = var_info['mip']
        logger.info(var_info['mip'])
        cmor_info = cmor_table.get_variable(var_info['mip'], var)
        _extract_variable(cmor_info, glob_attrs, filepath, out_dir)
```

Let's run this CMORizing script to see if the dataset is read correctly, and

> what kind of file is written out. Tere is a specific command available in the

Reviewer 2: ```suggestion
what kind of file is written out. There is a specific command available in the
```

Author: Done

ESMValTool to run the CMORizing scripts:

```bash
cmorize_obs -c <config-user.yml> -o <dataset-name>
```

The ``config-user-yml`` is the file in which we define the different data
paths, e.g. where the ESMValTool would find the "RAWOBS" folder. The
``dataset-name`` needs to be idential to the folder name that was created
to store the raw observation data files, in our case this would be "FLUXCOM".
The ESMValTool will create a folder with the correct tier information in your
defined output directory if that tier folder is not already available, and
then a folder named after the data set. In this folder the cmorized data set
will be stored as a netCDF file. If your run was successful, one or more
NetCDF files are produced in your output directory.

> ## Was the CMORization successful so far?!
>
> If you check the folders in your output path, you should see the following

> > folder structure:
> > ```bash
> > /Tier3/FLUXCOM/
> > ```

Reviewer 2: ```suggestion
> folder structure: ``/Tier3/FLUXCOM/``
```

Author: No longer part of the episode

>

> > Within the "FLUXCOM" folder there should be a NetCDF file with the name:
> > ```bash
> > OBS_FLUXCOM_reanaly_ANN-v1_Lmon_gpp_xxxx01-xxxx12.nc
> > ```

Reviewer 2: ```suggestion
> Within the "FLUXCOM" folder there should be a NetCDF file with the name:
> ``OBS_FLUXCOM_reanaly_ANN-v1_Lmon_gpp_xxxx01-xxxx12.nc``
```

Author: done

"time", "lat", "lon". The units for this variable are not defined yet.
The ESMValTool did not know how to convert from the
original units to the units that are defined in the CMOR table for the
variable "gpp". The original units are therefore listed in the "global
attributes" section as ``invalid_units``. The ESMValTool recognized that the
units given in the original file did not match the units given in the CMOR
table and therefore put the information about the units in the metadata.

If we test this NetCDF file now with our CMOR checker
```bash
esmvaltool run recipe_check_fluxcom.yml --log_level debug
```
we will encounter an error again. The dataset is not correctly CMORized, and
the first thing that the ESMValTool complains about is:
```bash
iris.exceptions.UnitConversionError: Cannot convert from unknown units. The
"units" attribute may be set directly.
```

Ok, so let's fix the units of the "GPP" variable in the CMORizer. For that
we add the following three lines to the code in the section
``_extract_variable``:

```python
# convert data from gc/m2/day to kg/m2/s
cube = cube / (1000 * 86400)
cube.units = 'kg m-2 s-1'
```

> The whole section should not look then look like this:

Reviewer 2: ```suggestion
The whole section should then look like this:
```

Author: No longer present

get a NetCDF file that has fixed units, but has an "unknown" variable name.

```bash
...
variables:
        float unknown(time, lat, lon) ;
                unknown:_FillValue = 1.e+20f ;
                unknown:units = "kg m-2 s-1" ;
        double time(time) ;
                time:axis = "T" ;
                time:units = "days since 1582-10-15 00:00:00" ;
                time:standard_name = "time" ;
                time:calendar = "gregorian" ;
        double lat(lat) ;
        double lon(lon) ;
...
```

We will have to do some more code tweaking then. But we also have not fixed
the problem with the coordinates ``lat`` and ``lon`` yet. There is no "units"
or "standard_name" given for either of these coordinates which will cause a
problem for the ESMValTool. Such some smaller formatting problems can occur
relatively often for coordinates like ``lat``, ``lon`` or ``time``. This means
that these problems need fixing in many CMORizers. Therefore there are common
functions available within the ESMValTool that one can import and use in the
new CMORizer script. The functions, written in python, are stored in the folder

> [utilities.py](https://github.com/ESMValGroup/ESMValTool/blob/master/esmvaltool/cmorizers/obs/utilities.py)

Reviewer 2: The script utilities.py has already been introduced above.

Author: Restructured so this is different now

>                 gpp:standard_name = "gross_primary_productivity_of_carbon" ;
>                 gpp:long_name = "Carbon Mass Flux out of Atmosphere due to Gross Primary Production on Land" ;
>                 gpp:units = "kg m-2 s-1" ;
>         double time(time) ;
>                 time:axis = "T" ;
>                 time:bounds = "time_bnds" ;
>                 time:units = "days since 1950-1-1 00:00:00" ;
>                 time:standard_name = "time" ;
>                 time:calendar = "gregorian" ;
>         double time_bnds(time, bnds) ;
>         double lat(lat) ;
>                 lat:axis = "Y" ;
>                 lat:bounds = "lat_bnds" ;
>                 lat:units = "degrees_north" ;
>                 lat:standard_name = "latitude" ;
>                 lat:long_name = "latitude coordinate" ;
>         double lat_bnds(lat, bnds) ;
>         double lon(lon) ;
>                 lon:axis = "X" ;
>                 lon:bounds = "lon_bnds" ;
>                 lon:units = "degrees_east" ;
>                 lon:standard_name = "longitude" ;
>                 lon:long_name = "longitude coordinate" ;
>         double lon_bnds(lon, bnds) ;
> ```
>

> > For that to happen, you will have to fix/work on the following things:

Reviewer 2: ```suggestion
> For that to happen, you have to fix/work on the following things:
```

Author: No longer present

> >                             unlimited_dimensions=['time'])
> >
> > ```
> >
> {: .solution}
{: .challenge}

So now we can finally run our new CMORizing script to produce an ESMValTool
readable datafile of the "FLUXCOM" dataset. As already mentioned, the call
to run a CMORizing script is as follows:

```bash
cmorize_obs -c <config-user.yml> -o <dataset-name>
```

If your run was successful, you should have gotten no error message in the
ESMValTool logging information and one or more NetCDF files should have been
produced in your output directory.

To make sure that the CMORization has really worked as planned, we run the
CMOR checker on this newly produced NetCDF file. If the ESMValTool can read
the data without problems, you should see as one of the last lines of
ESMValTool output:

```bash
INFO    Run was successful
```

## Last steps

Congratulations! You have successfully CMORized a new dataset!
Since you have gone through all the trouble to reformat the dataset so that
the ESMValTool can work with it, it would be great if you could provide the
CMORizer, and ultimately with that the dataset, to the rest of the community.
To do that there are a few more steps you have to do:

> 1. Open a pull request in the ESMValTool repository describing the dataset briefly

Reviewer 2: We encourage users to discuss their ideas in an issue as the first step of contribution. So, I suggest replacing this step with a sentence pointing to the episode 08-development-setup where the contribution procedure is explained.

Author: Done

2. Add the info of your dataset to the User Guide so that people know it is available for the ESMValTool [Obtaining input data](https://github.com/ESMValGroup/ESMValTool/blob/master/doc/sphinx/source/input.rst)
3. Make sure that there is a reference file available for the dataset [BibTeX info file](https://github.com/ESMValGroup/ESMValTool/tree/master/esmvaltool/references)

More information about working with pull requests are available in the ESMValTool

> documentation under [Contributing a review](https://esmvaltool--1920.org.readthedocs.build/en/1920/community/review.html)

Reviewer 2: ```suggestion
documentation under [Contributing a review](https://esmvaltool--1920.org.readthedocs.build/en/1920/community/review.html).
```

Author: Done

## Some final comments

Adding a new CMORizer to the ESMValTool is definitely already an advanced task

> when working with the ESMValTool. You have to have a basic understanding of

Reviewer 2: ```suggestion
when working with the ESMValTool. You need to have a basic understanding of
```

Author: Done

how the ESMValTool works and how it's internal structure looks like. In
addition, you need to have a basic understanding of NetCDF files and a
programming language. In our example we used python for the CMORizing script
since we advocate for focusing the code development on only a few different
programming languages. This helps to maintain the code and to ensure the
compatibility of the code with possible fundamental changes to the structure
of the ESMValTool and ESMValCore.

More information about adding observations to the ESMValTool can be found in the

> [documentation](https://docs.esmvaltool.org/en/latest/input.html#observations)

Reviewer 2: ```suggestion
[documentation](https://docs.esmvaltool.org/en/latest/input.html#observations).
```

Author: Done

> - standard_name: ``gross_primary_productivity_of_biomass_expressed_as_carbon``
> - frequency: ``mon``
> - modeling_realm: ``land``
>
> > ## Answers
> >
> > The variable "gpp" belongs to the land variables. The temporal resolution that we are looking
> > for is "monthly". This information points to the "Lmon" CMIP table. And indeed, the variable
> > "gpp" can be found in the file
> > [here](https://github.com/ESMValGroup/ESMValCore/blob/master/esmvalcore/cmor/tables/cmip6/Tables/CMIP6_Lmon.json).
> >
> {: .solution}
{: .challenge}

If the variable you are interested in is not available in the standard CMOR
tables, you could write a custom CMOR table entry for the variable. This,
however, is beyond the scope of this tutorial.

> ## Fill the configuration file
>
> Uncomment the following entries in your configuration file and fill them with appropriate values:
>
> - dataset_id
> - version
> - tier
> - modeling_realm
> - short_name (the ??? immediately under `variables`)
> - mip
>
> > ## Answers
> >
> > The configuration file now look something like this:
> >
> > ```yaml
> > ---
> > filename: 'GPP.ANN.CRUNCEPv6.monthly.*.nc'
> >
> > attributes:

> > >   project_id: OBS

Reviewer: Should be "OBS6" here, right?

Author: Well spotted! Changed it

> > ```
> >
> {: .solution}
{: .challenge}

Now that we have set this information correctly in the config file, we can call
the save function. Add the following python code to your CMORizer script:

```python
# 3. store the data with the correct filename
attributes = cfg['attributes']
variables = cfg['variables']

for short_name, variable_info in variables.items():
    all_attributes = {**attributes, **variable_info}  # add the mip to the other attributes
    utils.save_variable(cube=cube, var=short_name, outdir=out_dir, attrs=all_attributes)
```

Since we only have one variable (gpp), the loop is not strictly necessary.
However, this makes it possible to add more variables later on.

> ## Was the CMORization successful so far?
>
> If you run the CMORizer again, you should see that it creates an output file
> named ``OBS_FLUXCOM_reanaly_ANN-v1_Lmon_gpp_xxxx01-xxxx12.nc``. The "xxxx" and
> "yyyy" represent the start and end year of the data.
>
{: .callout}

Great! So we have produced a NetCDF file with the CMORizer that follows the
naming convention for ESMValTool datasets. Let's have a look at the NetCDF file
as it was written with the very basic CMORizer from above.

```bash

> ncdump -h OBS_FLUXCOM_reanaly_ANN-v1_Lmon_gpp_198001-198012.nc

Reviewer: And this should be for the year 2000 and not 1980. I had used 1980 as an example before I knew that you, Peter, had instructed people to download the year 2000.

Author: Check 👍

```

~~~
netcdf OBS_FLUXCOM_reanaly_ANN-v1_Lmon_gpp_198001-198012 {

> dimensions:

Reviewer: Also here: should be year 2000 and not 1980.

Author: Done

> >     CRUNCEP data file has several data files for different variables.
> >     The data for GPP generated using the
> >     Artificial Neural Network Method will be in files with name:
> >     GPP.ANN.CRUNCEPv6.monthly.\*.nc
> >     A registration is required for downloading the data.
> >     Users in the UK with a CEDA-JASMIN account may request access to the jules
> >     workspace and access the data.
> >     Note : This data may require rechunking of the netcdf files.
> >     This constraint will not exist once iris is updated to
> >     version 2.3.0 Aug 2019
> > """
> > ```
> {: .solution}
{: .challenge}

- **Complete the metadata in the config file**. We have left a few fields empty
  in the configuration file, such as 'source'. By filling out these fields we can
  make sure the relevant metadata is passed on as attributes in the CMORized
  data. To make this work, add the following line to the CMORizer script:

```python
# 2d. Update the cubes metadata with all info from the config file
utils.set_global_atts(cube, attributes)
```

- **Add a reference**. Make sure that there is a reference file available for

>    the dataset. You can add a [BibTeX info
>    file](https://github.com/ESMValGroup/ESMValTool/tree/master/esmvaltool/references)
>    here. You

Reviewer 2: ```suggestion
   the dataset, see the instruction [here](https://docs.esmvaltool.org/en/latest/community/diagnostic.html#adding-references).
```
